aws_lambda_powertools/utilities/idempotency/persistence/base.py

```diff
@@ -0,0 +1,441 @@
+        payload_validation_jmespath: str
+            A jmespath expression to extract the payload to be validated from the event record
+        expires_after_seconds: int
+            The number of seconds to wait before a record is expired
+        use_local_cache: bool, optional
+            Whether to locally cache idempotency results, by default False
+        local_cache_max_items: int, optional
+            Max number of items to store in local cache, by default 1024
+        hash_function: str, optional
+            Function to use for calculating hashes, by default md5.
+        """
+        self.event_key_jmespath = event_key_jmespath
+        if self.event_key_jmespath:
+            self.event_key_compiled_jmespath = jmespath.compile(event_key_jmespath)
+        self.expires_after_seconds = expires_after_seconds
+        self.use_local_cache = use_local_cache
+        if self.use_local_cache:
+            self._cache = LRUDict(max_items=local_cache_max_items)
+        self.payload_validation_enabled = False
+        if payload_validation_jmespath:
+            self.validation_key_jmespath = jmespath.compile(payload_validation_jmespath)
+            self.payload_validation_enabled = True
+        self.hash_function = getattr(hashlib, hash_function)
+
+    def _get_hashed_idempotency_key(self, lambda_event: Dict[str, Any]) -> str:
+        """
+        Extract data from lambda event using event key jmespath, and return a hashed representation
+
+        Parameters
+        ----------
+        lambda_event: Dict[str, Any]
+            Lambda event
+
+        Returns
+        -------
+        str
+            Hashed representation of the data extracted by the jmespath expression
+
+        """
+        data = lambda_event
+        if self.event_key_jmespath:
+            data = self.event_key_compiled_jmespath.search(lambda_event)
+        return self._generate_hash(data)
+
+    def _get_hashed_payload(self, lambda_event: Dict[str, Any]) -> str:
+        """
+        Extract data from lambda event using validation key jmespath, and return a hashed representation
+
+        Parameters
+        ----------
+        lambda_event: Dict[str, Any]
+            Lambda event
+
+        Returns
+        -------
+        str
+            Hashed representation of the data extracted by the jmespath expression
+
+        """
+        if not self.payload_validation_enabled:
+            return ""
+        data = self.validation_key_jmespath.search(lambda_event)
+        return self._generate_hash(data)
+
+    def _generate_hash(self, data: Any) -> str:
+        """
+        Generate a hash value from the provided data
+
+        Parameters
+        ----------
+        data: Any
+            The data to hash
+
+        Returns
+        -------
+        str
+            Hashed representation of the provided data
+
+        """
+        hashed_data = self.hash_function(json.dumps(data).encode())
```

Comment:
__Question__: This assumes that the data is always serializable. Are there any situation where an event payload or function response could contain non-serializable data?

E.g. `decimal.Decimal` is not serializable by default, but you can return a `decimal.Decimal` from a Lambda function.

Here's a Lambda function to extract how the Python runtime does it:

```python
def handler(event, context):
    with open("/var/runtime/lambda_runtime_marshaller.py") as f:
        print(f.read())
```

Reply:
Fixed - json serialization supports decimals now